IdentityCore/src/util/NSString+MSIDExtensions.m

```diff
@@ -95,26 +95,33 @@ - (NSString *)msidTrimmedString
     return [self stringByTrimmingCharactersInSet:set];
 }
 
+- (NSString *)msidURLDecode
+{
+    return CFBridgingRelease(CFURLCreateStringByReplacingPercentEscapes(kCFAllocatorDefault, (CFStringRef)self, CFSTR("")));
+}
 
 - (NSString *)msidWWWFormURLDecode
 {
     // Two step decode: first replace + with a space, then percent unescape
-    CFMutableStringRef decodedString = CFStringCreateMutableCopy( NULL, 0, (__bridge CFStringRef)self );
-    CFStringFindAndReplace( decodedString, CFSTR("+"), CFSTR(" "), CFRangeMake( 0, CFStringGetLength( decodedString ) ), kCFCompareCaseInsensitive );
-    
-    CFStringRef unescapedString = CFURLCreateStringByReplacingPercentEscapes( NULL,                    // Allocator
-                                                                                          decodedString,           // Original string
-                                                                                          CFSTR("")); // Encoding
-    CFRelease( decodedString );
-    
-    return CFBridgingRelease(unescapedString);
+    NSString *replacedString = [self stringByReplacingOccurrencesOfString:@"+" withString:@" "];
+    return [replacedString msidURLDecode];
 }
 
+- (NSString *)msidURLEncode
+{
+    return [[self msidURLEncodedIgnoreSpace] stringByReplacingOccurrencesOfString:@" " withString:@"%20"];
```

Comment:
Can't some kind of a system util take care of it, so we don't need to replace space with %20 manually?

Reply:
we can use stringByAddingPercentEncodingWithAllowedCharacters, but that'd end up doing percentage encoding twice.
if we really wanted to be strict, we could've done percentage encoding with defined character sets for each of the url components but here follows more "generic" encoding like previously.